jira/cookieauth.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+from __future__ import print_function
+
+"""
+This module implements cookie based authentication, which can be used to avoid
+a problem where JIRA can become unresponsive under great volume of HTTP basic
+auth requests (https://jira.atlassian.com/browse/JRASERVER-26397)
+"""
+
+import json
+import datetime
+
+import requests
+from requests.auth import AuthBase
+
+
+
+class JIRACookieAuth(AuthBase):
+    """ JIRA specific cookie authentication """
+
+    def __init__(self, username, password):
+        self.username = username
+        self.password = password
+        self.last_refresh = None
+        self.check_interval = datetime.timedelta(minutes=5)
+        self.cookies = None
+
+    def login(self, req):
+        server = req.url.replace(req.path_url, '')
+        login_url = "{}/rest/auth/1/session".format(server)
+        login_data = {'username': self.username, 'password': self.password}
+
+        session = requests.Session()
+        r = session.post(login_url,
+                         data=json.dumps(login_data),
+                         headers={'Content-Type': 'application/json'})
+
+        self.cookies = r.cookies
+        self.last_refresh = datetime.datetime.now()
```

Comment:
IMHO you should be using the session cookie expiry time to know when to refresh the cookie.